<a href="https://colab.research.google.com/github/mesushan/ANN-for-fashion-MNIST-dataset/blob/master/Building_an_Artificial_Neural_Network_in_TensorFlow_2_0_for_fashion_MNIST_dataset_with_parameter_tuning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data Source: [Kaggle](https://www.kaggle.com/zalando-research/fashionmnist)

## Step 1: Installing dependencies and setting up a GPU environment

In [0]:
!pip install tensorflow-gpu==2.1-rc0

## Step 2: Importing the libraries and dataset

In [0]:
import numpy as np
import datetime
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist

In [3]:
tf.__version__

'2.1.0-rc0'

## Step 3: Data Preprocessing

### Loading the dataset

In [4]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


### Normalizing the images

We divide each pixel of the image in the training and test sets by the maximum number of pixels (255).

In this way each pixel will be in the range [0, 1]. By normalizing images we make sure that our model (ANN) trains faster.

In [0]:
X_train = X_train / 255.0

In [0]:
X_test = X_test / 255.0

### Reshaping the dataset

Since we are building a fully connected network, we reshape the training set and the test set to be into the vector format.

In [0]:
# Since each image's dimension is 28x28, we reshape the full dataset to [-1 (all elements), height * width]
X_train = X_train.reshape(-1, 28*28)

In [8]:
X_train.shape

(60000, 784)

In [9]:
X_test.shape

(10000, 28, 28)

In [0]:
# We reshape the test set the same way as train set
X_test = X_test.reshape(-1, 28*28)

In [11]:
X_test.shape

(10000, 784)

## Step 4: Building an Artificial Neural Network

### Defining the model

We simply define an object of the Sequential model.

In [0]:
model = tf.keras.models.Sequential()

### Adding a first fully-connected hidden layer

Layer hyper-parameters to be used:
- number of units/neurons: 128
- activation function: ReLU
- input_shape: (784, )

In [0]:
model.add(tf.keras.layers.Dense(units=128, activation='relu', input_shape=(784, )))

### Adding the second hidden layer

Layer hyper-parameters to be used:
- number of units/neurons: 128
- activation function: ReLU

In [0]:
model.add(tf.keras.layers.Dense(units=24, activation='relu'))

### Adding a Dropout layer

Dropout is being used as it is a Regularization technique where we randomly set neurons in a layer to zero. That way while training those neurons won't be updated. Because some percentage of neurons won't be updated the whole training process is long and we have less chance for overfitting.

In [0]:
model.add(tf.keras.layers.Dropout(0.2))

### Adding the output layer

- units: number of classes (10 in the Fashion MNIST dataset)
- activation: softmax

In [0]:
model.add(tf.keras.layers.Dense(units=10, activation='softmax'))

### Compiling the model

- Optimizer: Adam
- Loss: Sparse softmax (categorical) crossentropy 

In [0]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['sparse_categorical_accuracy'])

In [18]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 24)                3096      
_________________________________________________________________
dropout (Dropout)            (None, 24)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                250       
Total params: 103,826
Trainable params: 103,826
Non-trainable params: 0
_________________________________________________________________


### Training the model

In [19]:
model.fit(X_train, y_train, epochs=10)

Train on 60000 samples
Epoch 1/10
60000/60000 [==============================] - 7s 123us/sample - loss: 0.6019 - sparse_categorical_accuracy: 0.7928
Epoch 2/10
60000/60000 [==============================] - 6s 98us/sample - loss: 0.4301 - sparse_categorical_accuracy: 0.8480
Epoch 3/10
60000/60000 [==============================] - 6s 102us/sample - loss: 0.3866 - sparse_categorical_accuracy: 0.8610
Epoch 4/10
60000/60000 [==============================] - 6s 99us/sample - loss: 0.3581 - sparse_categorical_accuracy: 0.8692
Epoch 5/10
60000/60000 [==============================] - 6s 96us/sample - loss: 0.3431 - sparse_categorical_accuracy: 0.8766
Epoch 6/10
60000/60000 [==============================] - 6s 95us/sample - loss: 0.3267 - sparse_categorical_accuracy: 0.8812
Epoch 7/10
60000/60000 [==============================] - 6s 94us/sample - loss: 0.3164 - sparse_categorical_accuracy: 0.8843
Epoch 8/10
60000/60000 [==============================] - 6s 94us/sample - loss: 0.3056 - spa

### Model evaluation and prediction

In [20]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)

10000/10000 [==============================] - 1s 84us/sample - loss: 0.3413 - sparse_categorical_accuracy: 0.8815


In [21]:
print("Test accuracy: {}".format(test_accuracy))

Test accuracy: 0.8815000057220459


From the above ANN, we see that our model makes the accuracy of 88.15% in test dataset which is almost similar to the last epoch in training set. With adding one hidden layer and increasing the eopch from 5 to 10, the predictions increase by approx 1% in both training and test.

From this two models we can see that with careful parameter tuning, better model can be made. If it was for the production purpose, this model can be further improved by trying different parameters tuning as done in this model.

